arrow-fx-kotlinx-coroutines/src/main/kotlin/arrow/integrations/kotlinx/extensions.kt

```diff
@@ -1,39 +1,105 @@
 package arrow.integrations.kotlinx
 
 import arrow.core.Either
+import arrow.fx.ForIO
+import arrow.fx.IO
 import arrow.fx.IOOf
 import arrow.fx.fix
+import arrow.fx.internal.UnsafePromise
+import arrow.fx.typeclasses.Disposable
+import arrow.fx.typeclasses.Fiber
 import kotlinx.coroutines.CancellationException
 import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.Job
+import kotlinx.coroutines.isActive
 import kotlinx.coroutines.newCoroutineContext
 import kotlinx.coroutines.suspendCancellableCoroutine
 import kotlin.coroutines.EmptyCoroutineContext
 import kotlin.coroutines.resume
 import kotlin.coroutines.resumeWithException
 
-suspend fun <A> IOOf<A>.suspendCancellable(): A = suspendCancellableCoroutine { cont ->
-  val disposable = fix().unsafeRunAsyncCancellable { result ->
-    result.fold<Unit>(cont::resumeWithException, cont::resume)
-  }
+/**
+ * Safely execute the source [IO] in a suspending context,
+ * doing so with [suspendCancellable] makes it participate in structured concurrency.
+ *
+ * This will make sure that the source [IO] is cancelled whenever it's [CoroutineScope] is cancelled.
+ */
+suspend fun <A> IOOf<A>.suspendCancellable(): A =
+  suspendCancellableCoroutine { cont ->
+    if (cont.isActive) {
```

Comment:
Nice, we should not even run our thing if the enclosing scope is already cancelled 👍